### PR TITLE
Bulk lookup for mobile users

### PR DESCRIPTION
**Pull request by orangejenny**
##### SUMMARY
https://dimagi-dev.atlassian.net/browse/ICDS-1199

##### FEATURE FLAG
"Bulk mobile worker management features: filtered download, bulk delete, and bulk lookup users."

This PR renames this flag, which used to be called "Ability to filter mobile workers based on role, location, and username when doing bulk download." It also deletes the "Allow bulk deletion of users based on a username upload." and moves that functionality under the same feature flag.

##### PRODUCT DESCRIPTION
Adds a "Bulk Lookup of Users" button to mobile workers page. @dimagi/product and @biyeun  I'm not proud of adding a sixth button to this row of buttons and am open to feedback. I considered putting tabs or panels on this page (one for users, one for "actions"?) but wasn't convinced it would be better.
Screen Shot 2020-03-09 at 7 11 38 PM

The upload page accepts a spreadsheet of usernames and then creates a download indicating which of those users are already present in the system.
Screen Shot 2020-03-09 at 7 12 08 PM

@AndreaMKing5 The feature flag is marked solutions limited, let me know if you would prefer something else. There's no harm or risk to this feature, I just don't imagine many projects having a use for it.

### CODE REVIEW

File: corehq/apps/users/views/mobile/users.py [hunk 1]
```diff
@@ -1226,6 +1241,43 @@ def _delete_users(self, request, user_docs_by_id, user_ids_with_forms):
...
+        if not usernames:
+            return self.get(request, *args, **kwargs)
+
+        user_docs_by_id = {doc['_id']: doc for doc in get_user_docs_by_username(usernames)}
+        usernames_not_found = self._get_usernames_not_found(request, user_docs_by_id, usernames, strict=False)
+
+        rows = []
+        for username in usernames:
+            rows.append([raw_username(username), _("no") if username in usernames_not_found else _("yes")])
+        writer.write([(tab_name, rows)])
```
Comment (mkangia) [inline corehq/apps/users/views/mobile/users.py]: nit: I think we can extract this logic for creating the excel sheet in a different method.

File: corehq/apps/users/views/mobile/users.py [hunk 2]
```diff
@@ -1129,7 +1132,55 @@ def get(self, request, domain, *args, **kwargs):
...
+            workbook = get_workbook(request.FILES.get('bulk_upload_file'))
+        except WorkbookJSONError as e:
+            messages.error(request, str(e))
+            return None
+
+        try:
+            sheet = workbook.get_worksheet()
+        except WorksheetNotFound:
+            messages.error(request, _("Workbook has no worksheets"))
+            return None
```
Comment (mkangia) [inline corehq/apps/users/views/mobile/users.py]: might be worth extracting _validate_workbook to put these two in there.

**Review by mkangia — APPROVED** [review]

File: corehq/toggles.py [hunk 3]
```diff
@@ -1354,6 +1354,15 @@ def _commtrackify(domain_name, toggle_is_enabled):
...
 
+BULK_USER_LOOKUP = StaticToggle(
+    'bulk_user_lookup',
+    "Bulk mobile worker lookup: upload a set of usernames and receive "
+    "a download noting which ones exist in the system.",
+    TAG_SOLUTIONS_LIMITED,
+    [NAMESPACE_DOMAIN],
+    help_link='https://confluence.dimagi.com/display/ccinternal/Bulk+Lookup+Users',
+)
+
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/toggles.py]: good idea, 132300ad5a4

[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/toggles.py]: good idea, 132300ad5a4

[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/toggles.py]: fyi @AndreaMKing5 the PR description is updated to reflect the changes in feature flags. There's no change for existing users; everyone who has the old bulk delete flag turned on also already had the filtered download flag turned on.

File: corehq/apps/users/views/mobile/users.py [hunk 4]
```diff
@@ -1226,6 +1241,43 @@ def _delete_users(self, request, user_docs_by_id, user_ids_with_forms):
...
+
+    def post(self, request, *args, **kwargs):
+        outfile = io.BytesIO()
+        writer = Excel2007ExportWriter()
+        tab_name = "users"
+        header_table = [(tab_name, [(_("username"), _("exists"))])]
+        writer.open(header_table=header_table, file=outfile)
+        usernames = self._get_usernames(request)
+        if not usernames:
+            return self.get(request, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: Reordered in f8e436eb8ab

File: corehq/apps/users/views/mobile/users.py [hunk 5]
```diff
@@ -1226,6 +1241,43 @@ def _delete_users(self, request, user_docs_by_id, user_ids_with_forms):
...
+
+        user_docs_by_id = {doc['_id']: doc for doc in get_user_docs_by_username(usernames)}
+        usernames_not_found = self._get_usernames_not_found(request, user_docs_by_id, usernames, strict=False)
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: Rearranged in 7db093cee94

File: corehq/apps/users/views/mobile/users.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: c644b441c75

File: corehq/apps/users/views/mobile/users.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: d41b116cac8

**Review by mkangia — APPROVED** [review]

**Comment by AndreaMKing5** [comment]
@orangejenny - based on [this page](https://www.commcarehq.org/hq/flags/), it looks like this is a Solutions-Open Use flag (not Limited Use), right?  Unless I'm looking at the wrong feature...

**Comment by orangejenny** [comment]
@AndreaMKing5 oops, yes. The flag I'd originally added was solutions limited, but the flag I ended up reusing is solutions open, I missed updating that part of the description.

**Pull request merged**